src/plugins/keras_support/categorical_crossentropy_operation.cpp

```diff
@@ -0,0 +1,291 @@
+// Copyright (c) 2018 Bita Hasheminezhad
+// Copyright (c) 2018 Hartmut Kaiser
+//
+// Distributed under the Boost Software License, Version 1.0. (See accompanying
+// file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+#include <phylanx/config.hpp>
+#include <phylanx/ir/node_data.hpp>
+#include <phylanx/execution_tree/primitives/node_data_helpers.hpp>
+#include <phylanx/plugins/keras_support/categorical_crossentropy_operation.hpp>
+#include <phylanx/util/blaze_traits.hpp>
+#include <phylanx/plugins/statistics/statistics_base.hpp>
+#include <phylanx/plugins/statistics/sum_operation.hpp>
+
+#include <hpx/include/lcos.hpp>
+#include <hpx/include/naming.hpp>
+#include <hpx/include/util.hpp>
+#include <hpx/throw_exception.hpp>
+#include <hpx/util/assert.hpp>
+
+#include <algorithm>
+#include <cstddef>
+#include <cstdint>
+#include <memory>
+#include <string>
+#include <utility>
+#include <vector>
+#include <type_traits>
+
+#include <blaze/Math.h>
+#if defined(PHYLANX_HAVE_BLAZE_TENSOR)
+#include <blaze_tensor/Math.h>
+#endif
+
+///////////////////////////////////////////////////////////////////////////////
+namespace phylanx { namespace execution_tree { namespace primitives
+{
+    ///////////////////////////////////////////////////////////////////////////
+    match_pattern_type const cat_cross_operation::match_data =
+    {
+        hpx::util::make_tuple("categorical_crossentropy",
+        std::vector<std::string>{
+            "categorical_crossentropy(_1,_2)",
+            "categorical_crossentropy(_1,_2,_3)"
```

Comment:
We can now support keyword and default arguments. In this case you could replace the two patterns you defined with
```
categorical_crossentropy(_1_target, _2_output, __arg(_3_from_logits, false))
```
This would a) assign names to the arguments, and b) provide a default argument value for the last argument. As long as this primitive is invoked through PhySL you can be sure it will receive 3 arguments.